core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala

```diff
@@ -694,7 +694,7 @@ private[storage] class PartiallyUnrolledIterator[T](
   }
 
   override def next(): T = {
-    if (unrolled == null) {
+    if (unrolled == null || !unrolled.hasNext) {
```

Comment:
> To throw a exception might be a little strange to me, as it still has elements so a no such element exception seems not correct. It just doesn't correctly fall back to rest.

I meant when `hasNext` returns `false`, `next` should return NoSuchElementException.